src/primitives/sinsemilla.rs

```diff
@@ -64,59 +61,124 @@ impl<I: Iterator<Item = bool>> Iterator for Pad<I> {
-///
-/// [concretesinsemillahash]: https://zips.z.cash/protocol/nu5.pdf#concretesinsemillahash
-pub(crate) fn hash(domain_prefix: &str, msg: impl Iterator<Item = bool>) -> pallas::Base {
-    extract_p(&hash_to_point(domain_prefix, msg))
+        padded
+            .chunks(K)
+            .fold(self.Q, |acc, chunk| acc.double() + S(chunk))
+    }
+
+    /// `SinsemillaHash` from [§ 5.4.1.9][concretesinsemillahash].
+    ///
+    /// [concretesinsemillahash]: https://zips.z.cash/protocol/nu5.pdf#concretesinsemillahash
+    pub(crate) fn hash(&self, msg: impl Iterator<Item = bool>) -> pallas::Base {
+        extract_p(&self.hash_to_point(msg))
+    }
+
+    /// Return `Q`
+    #[allow(non_snake_case)]
+    pub(crate) fn Q(&self) -> pallas::Point {
+        self.Q
+    }
 }
 
-/// `SinsemillaCommit` from [§ 5.4.7.4][concretesinsemillacommit].
+/// A domain in which [`sinsemilla::commit`] and
+/// [`sinsemilla::short_commit`] can be used.
 ///
-/// [concretesinsemillacommit]: https://zips.z.cash/protocol/nu5.pdf#concretesinsemillacommit
+/// [`sinsemilla::commit`]: self::commit
+/// [`sinsemilla::short_commit`]: self::short_commit
 #[allow(non_snake_case)]
-pub(crate) fn commit(
-    domain_prefix: &str,
-    msg: impl Iterator<Item = bool>,
-    r: &pallas::Scalar,
-) -> pallas::Point {
-    let m_prefix = domain_prefix.to_owned() + "-M";
-    let r_prefix = domain_prefix.to_owned() + "-r";
+pub struct CommitDomain {
+    Q: pallas::Point,
+    R: pallas::Point,
+}
 
-    let hasher_r = pallas::Point::hash_to_curve(&r_prefix);
+impl CommitDomain {
+    /// Construct a new CommitDomain with a specific prefix string
+    pub(crate) fn new(domain: &str) -> Self {
+        let m_prefix = domain.to_owned() + "-M";
+        let r_prefix = domain.to_owned() + "-r";
+        CommitDomain {
+            Q: HashDomain::new(&m_prefix).Q,
+            R: pallas::Point::hash_to_curve(&r_prefix.clone())(&[]),
+        }
+    }
 
-    hash_to_point(&m_prefix, msg) + hasher_r(&[]) * r
-}
+    /// `SinsemillaHashToPoint` from [§ 5.4.1.9][concretesinsemillahash].
+    ///
+    /// [concretesinsemillahash]: https://zips.z.cash/protocol/nu5.pdf#concretesinsemillahash
+    #[allow(non_snake_case)]
+    pub(crate) fn hash_to_point(&self, msg: impl Iterator<Item = bool>) -> pallas::Point {
+        let padded: Vec<_> = Pad::new(msg).collect();
 
-/// `SinsemillaShortCommit` from [§ 5.4.7.4][concretesinsemillacommit].
-///
-/// [concretesinsemillacommit]: https://zips.z.cash/protocol/nu5.pdf#concretesinsemillacommit
-pub(crate) fn short_commit(
-    domain_prefix: &str,
-    msg: impl Iterator<Item = bool>,
-    r: &pallas::Scalar,
-) -> pallas::Base {
-    extract_p(&commit(domain_prefix, msg, r))
+        let hasher_S = pallas::Point::hash_to_curve(S_PERSONALIZATION);
+        let S = |chunk: &[bool]| hasher_S(&lebs2ip_k(chunk).to_le_bytes());
+
+        padded
+            .chunks(K)
+            .fold(self.Q, |acc, chunk| acc.double() + S(chunk))
```

Comment:
Hmm, I do not like this duplication. Instead, store the `HashDomain` you are using to construct `Q`, and use that directly.